### DISCUSSION TITLE
Is the GDK suitable for Enterprise .Net developers for Gamepad Input

### DISCUSSION THREAD

**Discussion by ShadowMarker789** [microsoft/GDK#27] [Q&A]
[Microsoft Documentation](https://docs.microsoft.com/en-us/gaming/gdk/_content/gc/input/overviews/input-overview) says (Under Getting Started):

> GameInput is the recommended API for all new code, regardless of the target platform, because it provides support across all Microsoft platforms (including earlier versions of Windows), and provides superior performance versus legacy APIs.

And also says:

> GameInput is available on all Windows platforms—PC, Xbox, HoloLens, IoT, and others—and is callable from GDK, Win32, and Universal Windows Platform (UWP) applications. Most importantly, it's available on earlier versions of Windows all the way back to Windows 7 via a redistributable installer package. This enables a single input codebase to be used across the entire Microsoft ecosystem, with no platform/version special-casing required.

For enterprise developers using C# and the .Net toolings Is the GameInput APIs from the GDK suitable to use? 

We are migrating from .Net Framework v 4.5.2 to .Net 6 since v 4.5.2 ran out of support a while ago. 

In this upgrade process we were reviewing our APIs and was going to move from SharpDX.XInput which uses XInput v1.4 if available, falling back to v1.3, v9.1.0 sequentially if those are not available. 

Our use-case is embedded machine remote control using the gamepads to remotely control unmanned machinery. 

For .Net developers are we expecting there to be C# bindings available, and if not what's the modern recommended way for .Net applications to read input from controllers?

**Comment by BL1NDX3N0N** [comment]
There are major differences between XInput and newer input APIs. Newer input APIs are built specifically to handle features of gamepads following the release of the Xbox One. For example, haptics such as impulse triggers featured in Xbox One controllers are not available for interop in XInput. The same can be said for gamepad extensions such as keyboards which I am sure newer input APIs facilitate such functionality.

If you are not trying to target the Xbox One (and newer) or access the complete feature-set of controllers after the Xbox 360, which means you just want to poll for joystick and button input, then XInput will be more than sufficient for your use-cases. Anything more beyond that, you should be utilizing newer APIs Microsoft has to offer. I do want to mention that XInput isn't going anywhere, tons of modern games still utilize the API and it must exist for backwards compatibility but I cannot speak on behalf of Windows IoT. In regards to performance, you are not a user-group that is going to be taking advantage of such performance gains nor will they be noticeable for your current application, so it is negligeable. As far as newer APIs offering backwards compatibility, I heavily assume such is done by deferring calls to legacy APIs which include XInput and/or DirectInput.

If you are wanting a new wrapper that is more intuitive and documented for XInput than what you have now I seriously recommend you check out a wrapper I made over XInput 1.4 for those targeting .NET.

https://github.com/BlindEyeSoftworks/BlindEye.XInput

The licensing is pretty basic but if you do not like it I give you complete and express permission to apply your own license to it and use it however you like. If you need a custom build for your specific use-case I am also more than happy to write such free of charge if it's within scope. If there is anymore info you need in regards to XInput and newer APIs feel free to ask. I have updated my answer to correct mistakes and provide more insight, I will also be updating my answer once I have more information on the APIs available for those shipping code for Windows IoT.

EDIT:

After digging around through some source code and documentation it is becoming pretty apparent that the GameInput API is its own entirely brand new API serving as a replacement for XInput. Nowhere was I able to find, nor prove, that calls are deferred to legacy APIs such as XInput either, meaning they would not be shipped with Windows IoT nor would said drivers and legacy APIs be compatible. I do not know if you are wanting to use these gamepads wirelessly but if so it is going to be even more trickier since the only hardware Microsoft lists as compatible would be a USB Xbox 360 controller, meaning dongles would not be an option without creating your own driver. As far as the Xbox One's Bluetooth capable controllers, I am not seeing any drivers available for such. My recommendation is that depending on the amount of machines needing to be controlled you can setup a single server which you will be able to connect controllers to and address them to specific machines. That server would then consume input and send them over the network to applicable IoT devices. Microsoft has created a similar sample [here](https://docs.microsoft.com/en-us/samples/microsoft/windows-iotcore-samples/gopigoxbox/) doing such for controlling robots. But to answer your main question, aside from rolling your own drivers and an API, the GameInput API would be very well suited for your application since it is expected to be used in the AAA game-industry. Unless someone else is able to chime in, or another option does exist which I have overlooked or missed, it seems as though said API is really your only option for IoT aside from utilizing a server.

**Reply by ShadowMarker789** [reply]
[Re: Last Paragraph]
That is not too dissimilar to how we currently operate - we pipe over the current state of the entire input device (either X360-compat 
 gamepad, or other perhaps specialized input device) state over to the control-system after massaging it to map correctly to the PLC's native data-types. To pipe it over we use Allen Bradley EtherNet/IP (IP stands for Industrial Protocol, not Internet Protocol), since our PLCs are Allen Bradley Rockwell ControlLogix PLCs, typically L71 or L81 controllers. ControlLogix PLCs have different data-types, for instance there is no Float-64 data-type, and consecutive bools in a row get compressed down into a SINT [8-bit UINT] array to match, so we can't /just/ send over a C struct directly without being very very careful about it. 

I am not sure about wireless vs wired controllers. Some of our clients were using wireless controllers, but most of them have switched over to wired controllers for reasons unknown to me. 

We just mostly want to read the controllers/gamepads more easily. Our target operating system minimum versions for the computer(s) reading the input device is Windows 10 IoT 2019 LTSC, Windows Server 2019, Windows 10 1909 (That one might be revised later). That said, even Windows 10 IoT 2019 LTSC is built off of Windows Enterprise and should have/support the full set of windows features available on desktop machines. 

We are free to define the set of supported hardware, so this lets us potentially require clients to upgrade their controllers to ones that support GameInput if desirable, but it's possible that their controllers already support it. For instance, we recently shipped [ MODEL: FS-TBS-0700-01 SKU: 520557, Turtle Beach Recon Wired Controller ] which purports to be for Xbox One / Xbox Series X, Windows 10 PC. 

From our perspective as software developers, a single API for communicating with both standard gamepads and other input devices such as joysticks would be very valuable. But we're primarily C# developers, not that great with C/C++ . I am not very confident in my ability to create a wrapper for the GDK GameInput to expose this to C#/.Net. 

Given that the GameInput API is its own entirely new API serving as a replacement for older APIs, is there likely to be official wrappers for GameInput and similar APIs to be consumed by .Net developers, similar to the way [Windows.Gaming.Input](https://docs.microsoft.com/en-us/uwp/api/Windows.Gaming.Input?view=winrt-10240) functions? 

EDIT: 
I would typically walk the docs to make .Net/C# wrappers by finding the appropriate DLL entry-points for the target functions, but the [MS-Docs](https://docs.microsoft.com/en-us/gaming/gdk/_content/gc/reference/input/gameinput/interfaces/igameinput/methods/igameinput_getcurrentreading) have:

Header: GameInput.h

Library: xgameruntime.lib

Which is different to what I was expecting, I was expecting a DLL to dynamically link to, not a `.lib`.

**Reply by BL1NDX3N0N** [reply]
> That is not too dissimilar to how we currently operate - we pipe over the current state of the entire input device (either X360-compat
gamepad, or other perhaps specialized input device) state over to the control-system after massaging it to map correctly to the PLC's native data-types.

Honestly, that is probably the best way to do it unless you can implement a system with drivers readily available for it. Alternatively, Linux machines can be used and there are drivers and APIs out there for utilizing XInput devices.

> From our perspective as software developers, a single API for communicating with both standard gamepads and other input devices such as joysticks would be very valuable.

This is what Microsoft aims to do with the newer API it seems like.

> I am not very confident in my ability to create a wrapper for the GDK GameInput to expose this to C#/.Net.

No worries, wrapping unmanaged code is a pretty straight forward task once you get a hang of FFI, marshalling, etc. I am also quite certain Microsoft does have a repo available for generating managed wrappers over unmanaged code which could help automate things quite a bit but the scope of the newer API does not seem to be that big.

> Given that the GameInput API is its own entirely new API serving as a replacement for older APIs, is there likely to be official wrappers for GameInput and similar APIs to be consumed by .Net developers, similar to the way [Windows.Gaming.Input](https://docs.microsoft.com/en-us/uwp/api/Windows.Gaming.Input?view=winrt-10240) functions?

That I could not tell you, I assume that since Microsoft already created the API (albeit not entirely unmanaged) most people are going to stick with what they know and what Microsoft provides. I did some digging around trying to decompile some things and it appears as though there is no completely managed wrapper provided by Microsoft but instead the code is generated at compile time.

> Which is different to what I was expecting, I was expecting a DLL to dynamically link to, not a .lib.

You are not the only one caught by surprise, I too was expecting a native DLL that would be shipped as part of Windows for developers to reference or wrap which has certainly made giving you a clear cut answer quite difficult. It is also increasingly difficult to decompile sample projects and expecting managed code interop due to UWP and .NET Native. I haven't used UWP since Windows 8 so I am far from being an expert on it.

I am sorry about your current situation, I too was expecting a DLL that could just be simply wrapped and if that was the case (or if it is) I was more than happy to wrap it for you free of charge. Seems like you are stuck in a very uncommon situation and will either have to go with a platform fully supporting XInput or you are going to have to use the newer API. If going with the newer API you shouldn't have to wrap anything as it is part of the framework for UWP. I do not know if your company already pays Microsoft to access developer channels for support but if that is the case I would try reaching out to an engineer, they would have all the information you need. Quite unfortunate someone hasn't come in here already. If I find out anything more I will certainly let you know.

**Reply by ShadowMarker789** [reply]
I have made some progress. 

Expecting GameInput.dll to be in C:\\Windows\\System32\\GameInput.dll 

![image](https://user-images.githubusercontent.com/37165910/185940384-7d23de4f-0eac-4a7d-8fc2-3abd6afbcc4b.png)

You can load it and call. 

```cs
if(Environment.OSVersion.Platform == PlatformID.Win32NT)
{
	unsafe
	{
		Win32.HMODULE module;
		// Okay but do we have GameInput? 
		fixed (char* pGameInputDllName = "GameInput")
		{
			module = Win32.Windows.LoadLibraryW((ushort*)pGameInputDllName);
		}

		if (module == Win32.HMODULE.NULL || module == Win32.HMODULE.INVALID_VALUE)
		{
			Environment.FailFast("No GameInput DLL found");
		}
		string gameInputCreateName = "GameInputCreate";
		byte[] cstrGameInputCreateName = Encoding.UTF8.GetBytes(gameInputCreateName);
		IntPtr GameInputCreate = IntPtr.Zero;
		fixed (byte* pCurrentReadingName = cstrGameInputCreateName)
		{
			GameInputCreate = Win32.Windows.GetProcAddress(module, (sbyte*)pCurrentReadingName);
		}
		if (GameInputCreate == IntPtr.Zero)
		{
			Environment.FailFast("GameInput DLL does not have entrypoint called GameInputCreate");
		}

		var gameInputCreate = (delegate* unmanaged[Stdcall]<void*, Win32.HRESULT>)GameInputCreate;

		IntPtr testPointer;
		gameInputCreate(&testPointer);

		if (testPointer == IntPtr.Zero)
		{
			Environment.FailFast("Did not obtain a valid pointer from GameInputCreate");
		}
	}
}
```

Then there's a lot of Vtbl entries that you'll have to populate in C# to ensure that you're able to call everything correctly. 

I've had reasonable success so far. 

![image](https://user-images.githubusercontent.com/37165910/185940843-95d2c1e5-e955-4eba-97f6-971efdff672d.png)

With help from Tanner Gooding 's ClangSharp project, I was also able to process the GameInput.h file into a cs file that appears to work. The above screenshot used that. 

A single extern as below

```cs
[DllImport("GameInput.dll", CallingConvention = CallingConvention.StdCall, ExactSpelling = true)]
[return: NativeTypeName("HRESULT")]
public static extern int GameInputCreate(IGameInput** gameInput);
```

Everything else is various vtbl references. 

I could dump the cs file but it's 2319 lines long of various structs that have vtbl pointer members.

**Reply by ShadowMarker789** [reply]
![image](https://user-images.githubusercontent.com/37165910/185953927-429bc45e-0351-48c9-8cb7-bd7db826e931.png)

And got some readings, just the left-stick x, y displayed

**Reply by BL1NDX3N0N** [reply]
Very interesting. Did not know the library came pre-installed and thought it was something developers needed to redistribute themselves (hence the SDK). Once I have some time I will do some more RE work and testing on this. For the most part however, you shouldn't be attempting to wrap it yourself as Microsoft has already done that for you and you already have access to the API from a managed environment. Going this route is reinventing the wheel. I was decompiling only because I wanted to see if calls were being deferred to legacy APIs for backwards compatibility, I was not suggesting you consume the newer API yourself.

**Reply by ShadowMarker789** [reply]
> Very interesting. Did not know the library came pre-installed and thought it was something developers needed to redistribute themselves (hence the SDK). Once I have some time I will do some more RE work and testing on this. For the most part however, you shouldn't be attempting to wrap it yourself as Microsoft has already done that for you and you already have access to the API from an unmanaged environment. Going this route is reinventing the wheel. I was decompiling only because I wanted to see if calls were being deferred to legacy APIs for backwards compatibility, I was not suggesting you consume the newer API yourself.

There are redistributables so I hear. I have not yet found any. 🤷‍♀️ 

Also turns out many of our clients are still running Windows Server 2012 R2 which really dampens things as that one did NOT have GameInput preinstalled. 

I can always do runtime checks for API availability and polyfill from older APIs as necessary. 

E.g.: 
Prefer GameInput 
Else XInput 1.4 (Realistically everyone should have XInput 1.4)
Else XInput 1.3
Else XInput 0.9
Else DirectInput (Should never not be available)

And have a IInputProvider interface, and a set of classes, one class for each underlying implementation.

**Reply by BL1NDX3N0N** [reply]
I actually did some questioning around with a group of people I work with and asked them to look in their system directories for the DLL in order to confirm whether it ships with Windows, all of their systems (10 and 11) contained the DLL. So as far as redistributables go, seeing how it appears as though Windows ships with it, I don't exactly know what you would be redistributing unless said redistributables are intended for older versions of Windows which did not ship with it such as Windows 7 and 8. At this point it is just speculation because Microsoft is not being very open about this in their documentation and if they are being open then it is surely difficult to find information I may have overlooked.

I do not know if I would go to the extent of checking if the API is available however unless you plan on running this software across numerous versions of Windows. Even if such is the case, you are the one installing Windows and configuring the environment correct? If so, then it should always be available as you would be the one making sure it gets installed. I was also under the assumption this software was going to be exclusive to Windows IoT devices, if such is the case then the Game Input API is the only API available for said platform. Nowhere was I able to find documentation that indicates XInput and respective drivers are shipped or supported on Windows IoT.

Sorry for the small ramble, I have been up for two days straight assisting others and my brain is deprived right now. Need to get some sleep but please keep me updated! In the morning I'll send a message out to a group of people that are part of the Microsoft game dev community, they actively work with the GDK and Microsoft employees who also work with it tend to hang out there to help others. I think it is worth a shot.

**Reply by BL1NDX3N0N** [reply]
If there are any specific questions you would like me to ask them please note them here!

**Reply by ShadowMarker789** [reply]
> If there are any specific questions you would like me to ask them please note them here!

Yeah, mostly the redistributable packages, do they even exist? Are they in fact a thing? 

Regarding Windows IoT, I'm talking about the Windows 10 IoT 2019 LTSC and the Windows 10 IoT 2021 LTSC both of which are built off of an Enterprise version of windows, so it's fully-featured, full windows. Just doesn't have a bunch of preinstalled apps and whatnot. 

For us, we're running this on basically anything and everything. Some of our clients are running this on Windows Server 2012 R2, and are considering upgrading to Windows Server 2019, but I don't know if GameInput comes in Server 2019. 

I know that XInput works on the IoT devices we're running. One of the hardware devices is an AdLink MXE-5501 machine, which is quite a powerful industrial computer for its ability to survive 65 C ambient air temperature. 

Also we are NOT the ones installing windows sometimes. Sometimes it's done for us, and we just have to work with whatever virtualized server we get provided with. 🙃

So yeah, basically - Redistrutable for GameInput? and do they plan to make wrappers for the GDK/GameInput for us .Net developers? 

Thanks

**Comment by BL1NDX3N0N** [comment]
Sorry for the late reply, have been waiting for a response but have not been given one as the server has been pretty quiet lately. In my above comment about the DLL being shipped with Windows as well for the DLL being able to be redistributed, it should be pretty straight-forward to create your own bindings for the public API. To do such you can just use the public documentation for function names, parameters, and data types.

I am considering creating my own bindings like I did with XInput and releasing it but with the API still being relatively new I am going to wait as Microsoft may very well be working on one already. Each time I have tried to get ahead of Microsoft they have always released a similar project not even half-way through the course of my own which they also do a much better job at and have a much larger audience and reach.
